File: 01_source/wallet_service/app/schemas/wallet.py
from pydantic import BaseModel, Field


class BalanceOut(BaseModel):
    user_id: str
    balance: int
    version: int


class CreditBody(BaseModel):
    user_id: str
    amount: int = Field(gt=0)
    transaction_id: str


class DebitBody(BaseModel):
    user_id: str
    amount: int = Field(gt=0)
    transaction_id: str
    version: int | None = None


class ApplyCreditBody(BaseModel):
    user_id: str
    credit_id: str
    transaction_id: str


class ExpiredPickupBody(BaseModel):
    user_id: str
    amount: int = Field(gt=0)
    transaction_id: str

File: 01_source/frontend/src/api/wallet.ts
import { financeApi } from './financeClient'

const api = financeApi

export type BalanceOut = { user_id?: string; partner_id?: string; balance: number; version?: number }

export type TxRow = {
  id?: string
  transaction_id?: string
  type?: string
  status?: string
  amount?: number
  created_at?: string
  occurred_at?: string
}

export type DivergenceRow = {
  metric?: string
  divergence_percent?: number
  divergence_ratio?: number
  percent?: number
  details?: string
}

export type TransactionsFilter = {
  date_start?: string
  date_end?: string
  type?: 'credit' | 'debit' | ''
  status?: string
}

export const walletApi = {
  getBalance: (partnerId: string) => api.get<BalanceOut>(`/v1/wallet/${partnerId}/balance`),

  getTransactions: (partnerId: string, params?: TransactionsFilter) =>
    api.get<TxRow[]>(`/v1/wallet/${partnerId}/transactions`, { params }),

  getExpiredCredits: (partnerId: string) =>
    api.get<unknown>(`/v1/wallet/${partnerId}/credits/expired`).catch(() => ({ data: [] as unknown })),

  applyCredit: (partnerId: string, amount: number, orderId: string) =>
    api.post(`/v1/wallet/${partnerId}/apply-credit`, { amount, order_id: orderId }),

  reconcile: () => api.post(`/v1/wallet/reconcile`),

  getDivergences: () => api.get<DivergenceRow[]>(`/v1/wallet/divergences`),

  getReconcileHistory: () => api.get<unknown>(`/v1/wallet/reconcile/history`),
}
